components/alert/alert-toast.js

```diff
@@ -0,0 +1,183 @@
+			type: { type: String, reflect: true },
+			_state: { type: String, reflect: true }
+		};
+	}
+
+	static get styles() {
+		return css`
+			:host {
+				display: block;
+			}
+
+			.d2l-alert-toast-container {
+				border-radius: 0.3rem;
+				bottom: 1.5rem;
+				box-shadow: 0 0.1rem 0.6rem 0 rgba(0,0,0,0.10);
+				display: none;
+				left: 0;
+				margin: 0 auto;
+				max-width: 600px;
+				position: fixed;
+				right: 0;
+				width: 100%;
+				z-index: 10000;
+			}
+
+			:host(:not([_state="closed"])) .d2l-alert-toast-container {
+				display: block;
+			}
+
+			:host([_state="opening"]) .d2l-alert-toast-container,
+			:host([_state="closing"]) .d2l-alert-toast-container {
+				transition-duration: 250ms;
+				transition-property: transform, opacity, visibility;
+				transition-timing-function: ease-in;
+			}
+
+			:host([_state="preopening"]) .d2l-alert-toast-container,
+			:host([_state="closing"]) .d2l-alert-toast-container {
+				opacity: 0;
+				transform: translateY(0.5rem);
+				visibility: hidden;
+			}
+
+			:host([_state="opening"]) .d2l-alert-toast-container {
+				opacity: 1;
+				transform: translateY(0);
+				visibility: visible;
+			}
+
+			.d2l-alert-toast {
+				animation: none;
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this.hideCloseButton = false;
+		this.noAutoClose = false;
+		this.open = false;
+		this._state = states.CLOSED;
+	}
+
+	get _state() {
+		return this.__state;
+	}
+
+	set _state(val) {
+		const oldVal = this.__state;
+		if (oldVal !== val) {
+			this.__state = val;
+			this.requestUpdate('_state', oldVal);
+			this._stateChanged(val, oldVal);
+		}
+	}
+
+	get open() {
+		return this._open;
+	}
+
+	set open(val) {
+		const oldVal = this._open;
+		if (oldVal !== val) {
+			this._open = val;
+			this.requestUpdate('open', oldVal);
+			this._openChanged(val);
+		}
+	}
+
+	render() {
+		return html`
+			<div class="d2l-alert-toast-container" @transitionend=${this._onTransitionEnd}>
+				<d2l-alert class="d2l-alert-toast" type=${ifDefined(this.type)} @d2l-alert-closed=${this._onCloseClicked} button-text=${ifDefined(this.buttonText)} ?has-close-button=${!this.hideCloseButton} subtext=${ifDefined(this.subtext)}>
+					<slot></slot>
+				</d2l-alert>
+			</div>
+		`;
+	}
+
+	_onCloseClicked(e) {
+		e.target.removeAttribute('hidden');
+		this.open = false;
+	}
+
+	_onTransitionEnd() {
+		if (this._state === states.OPENING) {
+			this._state = states.OPEN;
+		} else if (this._state === states.CLOSING) {
+			this._state = states.CLOSED;
+		}
+	}
+
+	async _openChanged(newOpen) {
+		await this.updateComplete;
+
+		if (newOpen) {
+			if (this._state === states.CLOSING) {
+				this._state = states.OPENING;
+			} else if (this._state === states.CLOSED) {
+				if (!reduceMotion) {
+					this._state = states.PREOPENING;
+					// pause before running the opening animation because transitions won't run when changing from 'diplay: none' to 'display: block'
+					this._preopenFrame = requestAnimationFrame(() => {
```

Comment:
Ignoring the reduceMotion for the moment, would it be enough to:
```
this._state = states.PREOPENING;
await this.requestUpdate();
this._state = states.OPENING;
```
Or, if necessary to wait a frame, then:
```
this._state = states.PREOPENING;
await this.requestUpdate();
requestAnimationFrame(() => this._state = states.OPENING);
```
Lit will update the state as a micro-task.  Do we need to wait two frames?

Reply:
I played around with this a ton and the only thing that I found that worked is waiting two frames or longer. Waiting for the next frame after re-render unfortunately isn't enough.

Reply:
I've been playing around with this, and it is indeed a bit confusing.  I am aware of at least one bug (which may or may not be related) when setting secondary properties from a property setter, but I was still unable to get around this.  I did however find that this also works:
```
requestAnimationFrame(() => {
    this._state = states.PREOPENING;
    requestAnimationFrame(() => {
        this._state = states.OPENING;
    });
});
```
But is it still unclear what the root problem is, because prior to that code, there are no pending state changes that need to be rendered, painted, etc.

I also found that the `await this.updateComplete;` (line 140) is not necessary in this method... because there isn't anything in this method relies on previous changes being rendered, painted, etc.

Reply:
The `await this.updateComplete` is only unneeded if we move the `this._state = states.PREOPENING` into the `requestAnimationFrame`. I'm not entirely sure why but Lit seems to skip reflecting the `_state` if its updated in `_openChanged` without the `await this.updateComplete` causing the attribute go straight from `closed` to `opening`.

I want to avoid moving `states.PREOPENING` into the `requestAnimationFrame` because then there is a brief period where the component is in `CLOSED` state but is actually opening. This causes the component to end up open if `open` is quickly toggled from `false` to `true` then back to `false`.

Reply:
It looks like moving `state` to be on the `div` rather than reflected removes the need for `await this.updateComplete`. Almost feels like there is a Lit bug or gotcha I'm not aware of there related to reflected properties.